text/0037-node-upgrades.md

```diff
@@ -0,0 +1,95 @@
+# Node Upgrades
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0037](https://github.com/casperlabs/ceps/pull/0037)
+
+This CEP describes the how the `casper-node-launcher` and `casper-node` will interact to achieve network upgrades.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The `casper-node-launcher` is a relatively simple Rust binary which runs the `casper-node` binary as a child process.
+Normal operation will see the launcher running indefinitely, cycling upwards through versions of the node as and when
+they become active.
+
+The node has two relevant modes of operation: "validator mode", and "migrate-data mode".  The normal mode is validator.
+The migrate-data mode is used immediately after an upgrade to migrate any data stored on disk as required.
+
+The node also has three possible exit codes: 0, 101 or 102.  0 indicates success, 101 indicates an error (e.g. a panic)
+and 102 indicates the node should be downgraded to a previous version.  These exit codes are read and reacted to by the
+launcher.
+
+On startup, the launcher either tries to read its previously cached state from disk, or assumes a fresh start.  On a
```

Comment:
Can you include information about what kind of state it caches?

Reply:
Done in [b70fb7f](https://github.com/CasperLabs/ceps/pull/37/commits/b70fb7fee444b48e66112fda0a710b8c83a3ca85).